src/vt/vrt/collection/balance/temperedlb/temperedlb.cc

```diff
@@ -1296,14 +2146,586 @@ void TemperedLB::decide() {
+      this_new_work_, try_info.work, src_new_work, dest_new_work
+    );
+  };
+
+  // Identify and message beneficial cluster swaps
+  for (auto const& [try_rank, try_clusters] : other_rank_clusters_) {
+    bool found_potential_good_swap = false;
+
+    // Approximate the memory usage on the target
+    BytesType try_mem =
+      other_rank_working_bytes_.find(try_rank)->second;
+    for (auto const& [try_shared_id, try_cluster] : try_clusters) {
+      try_mem += try_cluster.bytes;
+    }
+
+    // Iterate over source clusters
+    for (auto const& [src_shared_id, src_cluster] : cur_clusters_) {
+      // Compute approximation swap criterion for empty cluster "swap" case
+      {
+        ClusterInfo empty_cluster;
+        double c_try = criterion(try_rank, try_mem, src_cluster, empty_cluster);
+        if (c_try > 0.0) {
+	  // Try to obtain lock for feasible swap
+          found_potential_good_swap = true;
+          proxy_[try_rank].template send<&TemperedLB::tryLock>(this_node, c_try);
+          break;
+        }
+      }
+
+      // Iterate over target clusters
+      for (auto const& [try_shared_id, try_cluster] : try_clusters) {
+	// Decide whether swap is beneficial
+        double c_try = criterion(try_rank, try_mem, src_cluster, try_cluster);
+        if (c_try > 0.0) {
+	  // Try to obtain lock for feasible swap
+          found_potential_good_swap = true;
+          proxy_[try_rank].template send<&TemperedLB::tryLock>(this_node, c_try);
+          break;
+        }
+      } // try_clusters
+      if (found_potential_good_swap) {
+        break;
+      }
+    } // cur_clusters_
+  } // other_rank_clusters
+
+  // We have to be very careful here since we will allow some reentrancy here.
+  constexpr int turn_scheduler_times = 10;
+  for (int i = 0; i < turn_scheduler_times; i++) {
+    theSched()->runSchedulerOnceImpl();
+  }
+
+  while (not theSched()->workQueueEmpty()) {
+    theSched()->runSchedulerOnceImpl();
+  }
+
+  ready_to_satisfy_locks_ = true;
+  satisfyLockRequest();
+
+  // Finalize epoch, we have sent our initial round of messages
+  // from here everything is message driven
+  theTerm()->finishedEpoch(lazy_epoch);
+  theTerm()->popEpoch(lazy_epoch);
+  vt::runSchedulerThrough(lazy_epoch);
+
+  vt_debug_print(
+    normal, temperedlb,
+    "After iteration: total memory usage={}, shared blocks here={}, "
+    "memory_threshold={}, load={}\n", computeMemoryUsage(),
+    getSharedBlocksHere().size(), mem_thresh_, this_new_load_
+  );
+
+  // Report on rejection rate in debug mode
+  if (theConfig()->vt_debug_temperedlb) {
+    int n_rejected = 0;
+    auto remote_block_count = getRemoteBlockCountHere();
+    runInEpochCollective("TemperedLB::swapClusters -> compute rejection", [=] {
+      proxy_.allreduce<&TemperedLB::rejectionStatsHandler, collective::PlusOp>(
+        n_rejected, n_transfers_swap_, remote_block_count
+      );
+    });
+  }
+} // void TemperedLB::originalTransfer()
```

Comment:
This is the end of `swapClusters`.
```suggestion
} // void TemperedLB::swapClusters()
```

Reply:
@cz4rs go ahead please then resolve thanks